File: Basics/Main/Bot.cs
using DisCatSharp.ApplicationCommands;
using DisCatSharp.ApplicationCommands.EventArgs;
using DisCatSharp.CommandsNext;
using DisCatSharp.Enums;
using DisCatSharp.Entities;
using DisCatSharp.EventArgs;
using DisCatSharp.Interactivity;
using DisCatSharp.Interactivity.Enums;
using DisCatSharp.Interactivity.Extensions;

using Microsoft.Extensions.Logging;

using System;
using System.Threading;
using System.Threading.Tasks;
using System.Reflection;

namespace DisCatSharp.Examples.Basics.Main;

/// <summary>
/// The bot.
/// </summary>
internal class Bot : IDisposable
{
#if DEBUG
	public static readonly string Prefix = "!";
#else
	public static readonly string Prefix = "%";
#endif
	//public static ulong devguild = ; //Set to register app command on guild

	public static CancellationTokenSource ShutdownRequest;
	public static DiscordClient Client;
	public static ApplicationCommandsExtension AppCommands;

	private InteractivityExtension _next;

	private CommandsNextExtension _cNext;

	/// <summary>
	/// Initializes a new instance of the <see cref="Bot"/> class.
	/// </summary>
	/// <param name="token">The token.</param>
	public Bot(string token)
	{
		ShutdownRequest = new();

		LogLevel logLevel;
#if DEBUG
		logLevel = LogLevel.Debug;
#else
		logLevel = LogLevel.Error;
#endif
		var cfg = new DiscordConfiguration
		{
			Token = token,
			TokenType = TokenType.Bot,
			AutoReconnect = true,
			MinimumLogLevel = logLevel,
			Intents = DiscordIntents.AllUnprivileged,
			MessageCacheSize = 2048
		};

		Client = new(cfg);

		Client.UseApplicationCommands();

		this._cNext = Client.UseCommandsNext(new()
		{
			StringPrefixes = [Prefix],
			CaseSensitive = true,
			EnableMentionPrefix = true,
			IgnoreExtraArguments = true,
			DefaultHelpChecks = null,
			EnableDefaultHelp = true,
			EnableDms = true
		});

		AppCommands = Client.GetApplicationCommands();

		this._next = Client.UseInteractivity(new()
		{
			PaginationBehaviour = PaginationBehaviour.WrapAround,
			PaginationDeletion = PaginationDeletion.DeleteMessage,
			PollBehaviour = PollBehaviour.DeleteEmojis,
			ButtonBehavior = ButtonPaginationBehavior.Disable
		});

		this.RegisterEventListener(Client, AppCommands, this._cNext);
		this.RegisterCommands(this._cNext, AppCommands);
	}

	/// <summary>
	/// Disposes the Bot.
	/// </summary>
	public void Dispose()
	{
		Client.Dispose();
		this._next = null;
		this._cNext = null;
		Client = null;
		AppCommands = null;
		Environment.Exit(0);
	}

	/// <summary>
	/// Starts the Bot.
	/// </summary>
	public async Task RunAsync()
	{
		await Client.ConnectAsync();
		while (!ShutdownRequest.IsCancellationRequested)
			await Task.Delay(2000);

		await Client.UpdateStatusAsync(null, UserStatus.Offline, null);
		await Client.DisconnectAsync();
		await Task.Delay(2500);
		this.Dispose();
	}

	/// <summary>
	/// Registers the event listener.
	/// </summary>
	/// <param name="client">The client.</param>
	/// <param name="cnext">The commandsnext extension.</param>
	private void RegisterEventListener(DiscordClient client, ApplicationCommandsExtension appCommands, CommandsNextExtension cnext)
	{
		/* Client Basic Events */
		client.SocketOpened += Client_SocketOpenedAsync;
		client.SocketClosed += Client_SocketClosedAsync;
		client.SocketErrored += Client_SocketErroredAsync;
		client.Heartbeated += Client_HeartbeatedAsync;
		client.Ready += Client_ReadyAsync;
		client.Resumed += Client_ResumedAsync;

		/* Client Events */
		//client.GuildUnavailable += Client_GuildUnavailable;
		//client.GuildAvailable += Client_GuildAvailable;

		/* CommandsNext Error */
		cnext.CommandErrored += CNext_CommandErroredAsync;

		/* Slash Infos */
		client.ApplicationCommandCreated += Discord_ApplicationCommandCreatedAsync;
		client.ApplicationCommandDeleted += Discord_ApplicationCommandDeletedAsync;
		client.ApplicationCommandUpdated += Discord_ApplicationCommandUpdatedAsync;
		appCommands.SlashCommandErrored += Slash_SlashCommandErroredAsync;
		appCommands.SlashCommandExecuted += Slash_SlashCommandExecutedAsync;
	}

	/// <summary>
	/// Registers the commands.
	/// </summary>
	/// <param name="cnext">The commandsnext extension.</param>
	/// <param name="appCommands">The appcommands extension.</param>
	private void RegisterCommands(CommandsNextExtension cnext, ApplicationCommandsExtension appCommands)
	{
		cnext.RegisterCommands<Commands.Main>(); // Commands.Main = Ordner.Class
		// appCommands.RegisterCommands<AppCommands.Main>(devguild); // use to register on guild
		appCommands.RegisterGlobalCommands<AppCommands.Main>(); // use to register global (can take up to an hour)
	}

	private static Task Client_ReadyAsync(DiscordClient dcl, ReadyEventArgs e)
	{
		Console.ForegroundColor = ConsoleColor.Yellow;
		Console.WriteLine($"Starting with Prefix {Prefix} :3");
		Console.WriteLine($"Starting {Client.CurrentUser.Username}");
		Console.WriteLine("Client ready!");
		Console.WriteLine($"Shard {dcl.ShardId}");
		Console.ForegroundColor = ConsoleColor.Yellow;
		Console.WriteLine("Loading Commands...");
		Console.ForegroundColor = ConsoleColor.Magenta;
		var commandlist = dcl.GetCommandsNext().RegisteredCommands;
		foreach (var command in commandlist)
			Console.WriteLine($"Command {command.Value.Name} loaded.");

		Console.ForegroundColor = ConsoleColor.Green;
		Console.WriteLine("Bot ready!");
		return Task.CompletedTask;
	}

	private static Task Client_ResumedAsync(DiscordClient dcl, ReadyEventArgs e)
	{
		Console.ForegroundColor = ConsoleColor.Yellow;
		Console.WriteLine("Bot resumed!");
		return Task.CompletedTask;
	}

	private static Task Discord_ApplicationCommandUpdatedAsync(DiscordClient sender, ApplicationCommandEventArgs e)
	{
		sender.Logger.LogInformation("Shard {SenderShardId} sent application command updated: {CommandName}: {CommandId} for {CommandApplicationId}", sender.ShardId, e.Command.Name, e.Command.Id, e.Command.ApplicationId);
		return Task.CompletedTask;
	}

	private static Task Discord_ApplicationCommandDeletedAsync(DiscordClient sender, ApplicationCommandEventArgs e)
	{
		sender.Logger.LogInformation("Shard {SenderShardId} sent application command deleted: {CommandName}: {CommandId} for {CommandApplicationId}", sender.ShardId, e.Command.Name, e.Command.Id, e.Command.ApplicationId);
		return Task.CompletedTask;
	}

	private static Task Discord_ApplicationCommandCreatedAsync(DiscordClient sender, ApplicationCommandEventArgs e)
	{
		sender.Logger.LogInformation("Shard {SenderShardId} sent application command created: {CommandName}: {CommandId} for {CommandApplicationId}", sender.ShardId, e.Command.Name, e.Command.Id, e.Command.ApplicationId);
		return Task.CompletedTask;
	}

	private static Task Slash_SlashCommandExecutedAsync(ApplicationCommandsExtension sender, SlashCommandExecutedEventArgs e)
	{
		sender.Client.Logger.LogInformation("Slash/Info: {ContextCommandName}", e.Context.CommandName);
		return Task.CompletedTask;
	}

	private static Task Slash_SlashCommandErroredAsync(ApplicationCommandsExtension sender, SlashCommandErrorEventArgs e)
	{
		sender.Client.Logger.LogError("Slash/Error: {ExceptionMessage} | CN: {ContextCommandName} | IID: {ContextInteractionId}", e.Exception.Message, e.Context.CommandName, e.Context.InteractionId);
		return Task.CompletedTask;
	}

	private static Task CNext_CommandErroredAsync(CommandsNextExtension sender, CommandErrorEventArgs e)
	{
		if (e.Command == null)
			sender.Client.Logger.LogInformation("{ExceptionMessage}", e.Exception.Message);
		else
			sender.Client.Logger.LogInformation("{CommandName}: {ExceptionMessage}", e.Command.Name, e.Exception.Message);

		return Task.CompletedTask;
	}

	private static Task Client_SocketOpenedAsync(DiscordClient dcl, SocketEventArgs e)
	{
		Console.ForegroundColor = ConsoleColor.Yellow;
		Console.WriteLine("Socket opened");
		return Task.CompletedTask;
	}

	private static Task Client_SocketErroredAsync(DiscordClient dcl, SocketErrorEventArgs e)
	{
		Console.ForegroundColor = ConsoleColor.DarkRed;
		Console.WriteLine("Socket has an error! " + e.Exception.Message.ToString());
		return Task.CompletedTask;
	}

	private static Task Client_SocketClosedAsync(DiscordClient dcl, SocketCloseEventArgs e)
	{
		Console.ForegroundColor = ConsoleColor.Red;
		Console.WriteLine("Socket closed: " + e.CloseMessage);
		return Task.CompletedTask;
	}

	private static Task Client_HeartbeatedAsync(DiscordClient dcl, HeartbeatEventArgs e)
	{
		Console.ForegroundColor = ConsoleColor.Yellow;
		Console.WriteLine("Received Heartbeat:" + e.Ping);
		Console.ForegroundColor = ConsoleColor.Gray;
		return Task.CompletedTask;
	}
}

File: FSharp/Basics/Bot.fs
module Basics.Bot

open System
open System.Threading
open System.Threading.Tasks
open Basics.AppCommands
open Basics.Commands
open DisCatSharp
open DisCatSharp.ApplicationCommands
open DisCatSharp.ApplicationCommands.EventArgs
open DisCatSharp.CommandsNext
open DisCatSharp.Common.Utilities
open DisCatSharp.Entities
open DisCatSharp.EventArgs
open DisCatSharp.Interactivity
open DisCatSharp.Interactivity.Enums
open DisCatSharp.Interactivity.Extensions
open Microsoft.Extensions.DependencyInjection
open Microsoft.Extensions.Logging

[<Literal>]
let prefix = "!"

type Bot =
    { Client: DiscordClient
      AppCommands: ApplicationCommandsExtension
      Interactivity: InteractivityExtension
      CommandsNext: CommandsNextExtension
      ShutdownRequest: CancellationTokenSource }
    interface IDisposable with
        member this.Dispose() =
            this.ShutdownRequest.Dispose()
            this.Client.Dispose()

type private ClientHandler<'a when 'a :> AsyncEventArgs> = AsyncEventHandler<DiscordClient, 'a>

/// <summary>
/// Perform an action and return a task encapsulating it.
/// </summary>
/// <param name="f">The action to perform.</param>
/// <param name="m">The object to apply to the action.</param>
let inline private taskR f m = task { f m } :> Task

[<RequireQualifiedAccess>]
module rec Bot =
    /// <summary>
    /// Creates an instance of the <see cref="Bot"/> record.
    /// </summary>
    /// <param name="token">The Discord token to use in configuration.</param>
    let create token =
        let shutdownRequest = new CancellationTokenSource()
        // Create an IServiceProvider with the shutdown request so it can be
        // used by commands.
        let services =
            ServiceCollection()
                .AddSingleton<CancellationTokenSource>(shutdownRequest)
                .BuildServiceProvider()

        let logLevel =
#if DEBUG
            LogLevel.Debug
#else
            LogLevel.Error
#endif
        let cfg = DiscordConfiguration()
        cfg.Token <- token
        cfg.TokenType <- TokenType.Bot
        cfg.AutoReconnect <- true
        cfg.MinimumLogLevel <- logLevel
        cfg.Intents <- DiscordIntents.AllUnprivileged
        cfg.MessageCacheSize <- 2048

        let client = new DiscordClient(cfg)

        let commandsNextCfg = CommandsNextConfiguration()
        commandsNextCfg.StringPrefixes <- [| prefix |]
        commandsNextCfg.CaseSensitive <- true
        commandsNextCfg.EnableMentionPrefix <- true
        commandsNextCfg.IgnoreExtraArguments <- true
        commandsNextCfg.DefaultHelpChecks <- null
        commandsNextCfg.EnableDefaultHelp <- true
        commandsNextCfg.EnableDms <- true
        let commandsNext = client.UseCommandsNext commandsNextCfg

        let appCommandsCfg = ApplicationCommandsConfiguration()
        appCommandsCfg.Services <- services
        let appCommands = client.UseApplicationCommands()

        let interactivityCfg = InteractivityConfiguration()
        interactivityCfg.PaginationBehaviour <- PaginationBehaviour.WrapAround
        interactivityCfg.PaginationDeletion <- PaginationDeletion.DeleteMessage
        interactivityCfg.PollBehaviour <- PollBehaviour.DeleteEmojis
        interactivityCfg.ButtonBehavior <- ButtonPaginationBehavior.Disable
        let interactivity = client.UseInteractivity interactivityCfg

        let bot =
            { Client = client
              AppCommands = appCommands
              Interactivity = interactivity
              CommandsNext = commandsNext
              ShutdownRequest = shutdownRequest }

        registerEventListener bot
        registerCommands bot
        bot

    let runAsync bot =
        task {
            do! bot.Client.ConnectAsync()

            while not <| bot.ShutdownRequest.IsCancellationRequested do
                do! Task.Delay 2000

            do! bot.Client.UpdateStatusAsync(null, UserStatus.Offline, Nullable())
            do! bot.Client.DisconnectAsync()
            do! Task.Delay 2500
            (bot :> IDisposable).Dispose()
        }

    let registerEventListener
        { Client = client
          AppCommands = appCommands
          CommandsNext = commandsNext }
        =
        // Basic client events
        let client_SocketOpened =
            ClientHandler<SocketEventArgs>(fun c _ -> taskR c.Logger.LogInformation "Socket opened")

        let client_SocketErrored =
            ClientHandler<SocketErrorEventArgs>
                (fun c e -> taskR c.Logger.LogError $"Socket has an error! %s{e.Exception.Message}")

        let client_SocketClosed =
            ClientHandler<SocketCloseEventArgs>
                (fun c e -> taskR c.Logger.LogInformation $"Socket closed: %s{e.CloseMessage}")

        let client_Heartbeated =
            ClientHandler<HeartbeatEventArgs>(fun c e -> taskR c.Logger.LogDebug $"Received heartbeat: %i{e.Ping}")

        let client_Ready =
            let handler (client: DiscordClient) (args: ReadyEventArgs) =
                seq {
                    yield $"Starting with prefix %s{prefix}"
                    yield $"Starting %s{client.CurrentUser.Username}#%s{client.CurrentUser.Discriminator}"
                    yield "Client ready!"
                    yield $"Shard: %u{client.ShardId}"
                    yield "Loaded commands:"

                    yield!
                        client.GetCommandsNext().RegisteredCommands
                        |> Seq.map (fun kvp -> $"    %s{kvp.Value.Name}")

                    yield "Bot ready!"
                }
                |> taskR (Seq.iter client.Logger.LogInformation)

            ClientHandler<ReadyEventArgs>(handler)

        let client_Resumed =
            ClientHandler<ReadyEventArgs>(fun c _ -> taskR c.Logger.LogInformation "Bot resumed!")

        client.add_SocketOpened client_SocketOpened
        client.add_SocketErrored client_SocketErrored
        client.add_SocketClosed client_SocketClosed
        client.add_Heartbeated client_Heartbeated
        client.add_Ready client_Ready
        client.add_Resumed client_Resumed

        // Commands next
        let commandsNext_CommandErrored =
            AsyncEventHandler<CommandsNextExtension, CommandErrorEventArgs>
                (fun c e ->
                    match e.Command with
                    | null -> e.Exception.Message
                    | _ -> $"%s{e.Command.Name}: %s{e.Exception.Message}"
                    |> taskR c.Client.Logger.LogError)

        commandsNext.add_CommandErrored commandsNext_CommandErrored

        // Application commands
        let discord_ApplicationCommandCreated =
            ClientHandler<ApplicationCommandEventArgs>
                (fun c e ->
                    taskR
                        c.Logger.LogInformation
                        $"Shard %u{c.ShardId} sent application command created: %s{e.Command.Name}: (%u{e.Command.Id}) for %u{e.Command.ApplicationId}")

        let discord_ApplicationCommandDeleted =
            ClientHandler<ApplicationCommandEventArgs>(fun c e -> taskR c.Logger.LogInformation $"")

        let discord_ApplicationCommandUpdated =
            ClientHandler<ApplicationCommandEventArgs>(fun c e -> taskR c.Logger.LogInformation $"")

        let slash_SlashCommandErrored =
            AsyncEventHandler<ApplicationCommandsExtension, SlashCommandErrorEventArgs>
                (fun c e -> taskR c.Client.Logger.LogError $"")

        let slash_SlashCommandExecuted =
            AsyncEventHandler<ApplicationCommandsExtension, SlashCommandExecutedEventArgs>
                (fun c e -> taskR c.Client.Logger.LogError $"")

        client.add_ApplicationCommandCreated discord_ApplicationCommandCreated
        client.add_ApplicationCommandDeleted discord_ApplicationCommandDeleted
        client.add_ApplicationCommandUpdated discord_ApplicationCommandUpdated

        appCommands.add_SlashCommandErrored slash_SlashCommandErrored
        appCommands.add_SlashCommandExecuted slash_SlashCommandExecuted

        ()

    let registerCommands
        { CommandsNext = commandsNext
          AppCommands = appCommands }
        =
        commandsNext.RegisterCommands<CommandsNext.Main>()
        appCommands.RegisterCommands<AppCommands.Main>()

        ()
